### PR TITLE
Try Blazor AOT after replacing Blazor Worker

### PR DESCRIPTION
**Pull request by StachuDotNet**
## Phase 1: make blazor-loading simpler
- [x] Build an un-minified copy of `blazor.webassembly.js` from the the [aspnetcore repo](https://github.com/dotnet/aspnetcore)
  this took a `./restore.sh` and `./build.sh` in the `src/Components` directory, after adjusting `Components/Web.JS/src/webpack.config.js` such that `minimize: false` under the `optimization` option
- [x] Copy that into Dark (fsharp-backend/src/Wasm/static` as `blazor.webassembly.js`)
- [x] Prepare a custom resource-loader to respect Dark having files at `{static host}/blazor/*` rather than `.../_framework/*`
  I tossed this into `appsupport.js` for now
- [x] Reference the new `blazor.webassembly.js` in `ui.html`
  `autostart="false"` is needed so `appsupport` has a chance to avail some relevant pieces
- [ ] With that in place, load Blazor resources fully, ready for interop
  I'm loading this with `Blazor.start({ loadBootResource: loadBootResource })`
  Most resources (boot.js and all the .dlls) are loading again well.
  The next failure type (not currently blocked - just pausing for a break now) is on `dotnet.wasm` - for some reason it's trying to locate at `http://darklang.localhost:9000/a/dotnet.wasm` rather than using the normal resource-loader. Maybe we need to pass something else into `Blazor.start`

## Phase 2: Find the minimal bug relevant to AOT builds
Once the simpler blazor-loading implementation is in place, the goal is to get to _some_ working build using AOT compiles, and a minimal step required to _break_ that

- [ ] Gut as much of the Wasm-build Dark as possible (e.g. most fns)
- [ ] Get basic calls working
- [ ] Add as much back in as we can, excluding anything that breaks
- [ ] Collect a report on the minimal breakage
- [ ] Figure out what to do from there

Misc. notes and TODOs:

- For now I'm not trying to load both the old-style BlazorWorker and the more 'raw' version - too much distracting network traffic
-  blazor.webassembly.js has changed a bit since we last copied it, so I opted to grab a fresher copy, more relevant for conversations with aspnetcore folks
- [ ] Review #3642 changes; there may be something here worth salvaging.
- [ ] Find/create an issue in `aspnetcore` of `Blazor.start` not being documented
- [ ] Publish npm package for `blazor.webassembly.js`, and document? create bindings (.ts, .fs, rescript)

### CODE REVIEW

**Comment by StachuDotNet** [comment]
Still haven't figured out why `dotnet.wasm` is failing to download.
image

The file is requested multiple times:
- once at the correct URL (`http://static.darklang.localhost:9000//blazor/dotnet.wasm`), which 200s
- twice at `http://darklang.localhost:9000/a/dotnet.wasm`, which 500s
- each of those 500s is somehow paired with this console error message:
  `Unhandled promise rejection unhandledrejection RuntimeError: abort(RuntimeError: abort(both async and sync fetching of the wasm failed). Build with -s ASSERTIONS=1 for more info.). Build with -s ASSERTIONS=1 for more info.`
- searching for that text, I find it (only) in 3 files (one relevant):
  ```
  dark@dark-dev:~/app$ grep -rnw '.' -e 'Build with -s ASSERTIONS=1'
  ./backend/static/blazor/dotnet.6.0.4.dze57nx010.js:1:...
  ./fsharp-backend/Build/out/Wasm/Debug/net6.0/dotnet.js:1:...
  ./fsharp-backend/Build/out/Wasm/Debug/net6.0/wwwroot/_framework/dotnet.6.0.4.dze57nx010.js:1:...
  ```

Something is loading dotnet.wasm, outside of blazor.webassembly.js and it seems to break the loading.

Some questions to pursue:
- [ ] review the above screen-shot some more; I haven't looked too deeply yet
- [ ] what else is loading this file?
- [ ] why is it doing so _twice_?
- [ ] how/can I get it to load via our custom resource-loader?
- [ ] is the problem in `dotnet.js`, or is it something in blazor.webassembly.js?
- [ ] maybe it's worth reading through blazor.webassembly.js more to figure out how/where it I/O's with blazor.js

**Comment by pbiggar** [comment]
> The next failure type (not currently blocked - just pausing for a break now) is on dotnet.wasm - for some reason it's trying to locate at http://darklang.localhost:9000/a/dotnet.wasm rather than using the normal resource-loader. Maybe we need to pass something else into Blazor.start

This script is being run at `darklang.localhost/a/canvasname`, while the old script ran in a web worker, so presumably it had to http path to be relative to. There's a function somewhere where we translate the blazor path into a url, you can probably use that to solve this.

**Comment by StachuDotNet** [comment]
If you're referring to `toDarklangSetupUri`, that's already been ported, used in the code here, and respected in most of the requests. The issue is that there are _extra_ calls which don't seem to be flowing through the `loadBootResource` param passed into `Blazor.start`. (our argument of `loadBootResource` includes flowing through `toDarklangSetupUri`)

**Comment by StachuDotNet** [comment]
- [ ] actually follow `Build with -s ASSERTIONS=1 for more info.`
  (I'm assuming this is a flag to set in the wasm.fsproj - haven't looked yet)

**Comment by StachuDotNet** [comment]
I just added a _rough_ hack to proceed - added a new route in apiserver to redirect the attempted /a/dotnet.wasm requests to the static host. It seems to get me past that step. Something to follow up on.

File: fsharp-backend/src/ApiServer/ApiServer.fs [hunk 1]
```diff
@@ -68,6 +68,17 @@ let addRoutes
   let std = standardMiddleware
   let html = htmlMiddleware
 
+  // this is a bad hack since I can't seem to adjust dotnet.js to load dotnet.wasm anywhere else...
+  // todo: load the real dotnet.wasm here... using httpclient? gross.
+  // todo: a 301 would be a bit less hacky - but would it work?
```
Comment (StachuDotNet) [inline fsharp-backend/src/ApiServer/ApiServer.fs:73]: it does.

**Comment by StachuDotNet** [comment]
Blazor.start() is now executing without error, but after throwing in many console.logs, I've seen that we're getting stuck in an `createEmscriptenModuleInstance` fn - trying to diagnose why.

**Comment by StachuDotNet** [comment]
I spent a while trying to get this (using blazor.webassembly.js rather than our BlazorWorker) working with .NET 6.
It kept getting stuck, with no console errors.

Then, I tried the same approach (using blazor.webassembly.js rather than BlazorWorker) against the .NET 7 branch, and I'm able to get basic evals working. (without AOT - about to try such)

I'm guessing that the issue here is that I was using a more 'modern' blazor.webassembly.js, incompatible with .net 6 for some reason. I'm going to continue working on the .net7 branch tonight, but plan to follow up and generate a raw .net 6-friendly blazor.webassembly.js tomorrow if possible. (by checking out `aspnetcore` at an earlier point).

Happy to finally get evals working with blazor.webassembly.js rather than BlazorWorker, now that it's finally loading fully rather than getting stuck.

Now, I'm going to try AOT builds with this set-up, and go from there.

**Comment by StachuDotNet** [comment]
When attempting to run against AOT builds (using the simplified blazor-loading in .net 7 branch), I'm coming across this error:
image

`dotnet.js` is looking for `Microsoft.AspNetCore.Components.WebAssembly`, but
- it doesn't exist in the `publish`ed build, so doesn't end up in `backend/static/blazor`
- because of this, it also isn't listed in `blazor.boot.js`, which `dotnet.js` uses to try to find the appropriate `Microsoft.AspNetCore.Components.WebAssembly` dll

Why do some .dlls simply not exist in the `publish` directory? This is a fundamental question that I'm not getting. I can't figure out why `Microsoft.AspNetCore.Components.WebAssembly` wouldn't be published with the other .dlls

**Comment by StachuDotNet** [comment]
Sneaking in `Microsoft.AspNetCore.Components.WebAssembly` (copying the .dll from non-publish artifacts, and adding the relevant line to blazor.boot.js) got a bit further, but failed shortly after.
image.

Not sure.

**Comment by StachuDotNet** [comment]
https://github.com/fsprojects/Paket/issues/3999 This may be relevant to our issue - trying the workaround mentioned in comments.

**Comment by StachuDotNet** [comment]
Made some great Blazor discoveries yesterday. Following up on them today.

I need to expand upon these notes more, and push my local code[1], but here are some take-aways:
- After varied adjustments, I'm able to load and run the AOT build with the `blazor.webassembly.js` script, rather than our hack
  Note: This testing has been _without_ a web worker. Will need to follow up to not block the UI thread
- My to-go test case is simple (`List::repeat 10000 1 |> List::length`), but it shows clear differences between AOT and non-AOT results, so I'm happy with it for now
- The root cause of perf, for my testing, is: serialization/deserialiation.
- Specifically, my test's results include some intermediate results for the `List::repeat` - that long list of `1`s is taking ages to serialize
- The serialization of these 10k elements takes ~13s with a non-AOT build, and ~9s with an AOT build. While this isn't a huge improvement, it's great that:
  1. the AOT build is working
  2. we see actual improvement with the AOT build
  3. seemingly, we've centralized the problem to serialization (again, at least for my cases)

[1] my local code is in a bit of a mess now; for now I've pushed the mess to [another branch](https://github.com/StachuDotNet/dark/commit/b911f4c1a2ab1f2ae790f532c2325217c31b0885) - currently tidying and including here.

**Comment by StachuDotNet** [comment]
Some additional notes:
- I had to do things to our Wasm.fsproj to get the AOT build to work, because of some bug in Paket, I believe. This might be fine for now, but we're including a lot of stuff that doesn't really matter right now - I think we can trim a lot more with a bit more research. (and maybe a contribution to Paket)
- the mono console warnings that I saw before in trying to use AOT builds seemingly relate to tasks - I had to update some things in Wasm.fs to get around this, though I'm hoping to simplify and have more answers here

**Comment by StachuDotNet** [comment]
@pbiggar you mentioned hitting perf issues on various canvases before while just poking around. My guess is that those perf issues were _also_ due to the serialization bottleneck - if you happen to recall if those handlers would involve large[1] inputs, outputs, or intermediate results, that'd be promising context to have.

[1] "large" here means "involves many elements" such that serialization may take a while.

**Comment by StachuDotNet** [comment]
Closing in favor of tidier PRs

**Pull request closed without merging**